lib/auth.js

```diff
@@ -0,0 +1,189 @@
+/*!
+ * Tozny e3db
+ *
+ * LICENSE
+ *
+ * Tozny dual licenses this product. For commercial use, please contact
+ * info@tozny.com. For non-commercial use, the contents of this file are
+ * subject to the TOZNY NON-COMMERCIAL LICENSE (the "License") which
+ * permits use of the software only by government agencies, schools,
+ * universities, non-profit organizations or individuals on projects that
+ * do not receive external funding other than government research grants
+ * and contracts.  Any other use requires a commercial license. You may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at https://tozny.com/legal/non-commercial-license.
+ * Software distributed under the License is distributed on an "AS IS"
+ * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
+ * License for the specific language governing rights and limitations under
+ * the License. Portions of the software are Copyright (c) TOZNY LLC, 2017.
+ * All rights reserved.
+ *
+ * @copyright Copyright (c) 2017 Tozny, LLC (https://tozny.com)
+ */
+
+/* global fetch */
+
+import 'isomorphic-fetch'
+
+import { default as uuid } from 'uuid/v4'
+
+const supportedTSV1Algorithms = ['TSV1-ED25519-BLAKE2B']
+
+/**
+ * Fallback polyfill to allow for HTTP Basic authentication from either Node
+ * or browser-based JavaScript.
+ *
+ * @param {string} str String to encode as Base64
+ */
+function btoa(str) {
+  let buf = Buffer.from(str, 'utf8')
+  return buf.toString('base64')
+}
+
+function sortQueryParams(queryString) {
+  // Parse query string into field=values
+  var query = queryString.split('&')
+  for (var i = 0; i < query.length; i++) {
+    // Separate fields from values
+    let splitQuery = query[i].split('=')
+    query[i] = splitQuery
+  }
+  // Sort alphabetically
+  query.sort(function(a, b) {
+    return a[0] < b[0] ? -1 : 1
+  })
+  // Reconstruct query
+  for (var j = 0; j < query.length; j++) {
+    query[j] = query[j].join('=')
+  }
+  return query.join('&')
+}
+
+class Auth {
+  constructor(apiKeyId, apiSecret, publicSigningKey, privateSigningKey, crypto, apiUrl) {
+    this.apiKeyId = apiKeyId
```

Comment:
Not convinced I like this authenticated request system to be stateful, but I don't have anything against it either, and it allows it to abstract token caching away from actual API business logic.

One change I would very much like, though, would be to take an options object instead of parameterized values. This will allows us to add/subtract supported items more flexibly going forward, as well as pass fewer null/undefined values into unused properties.